File: samples/python/src/roles/merchant_payment_processor_agent/tools.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Tools for the merchant payment processor agent.

Each agent uses individual tools to handle distinct tasks throughout the
shopping and purchasing process.
"""


import logging
from typing import Any

from a2a.server.tasks.task_updater import TaskUpdater
from a2a.types import DataPart
from a2a.types import Part
from a2a.types import Task
from a2a.types import TaskState
from a2a.types import TextPart

from ap2.types.mandate import PAYMENT_MANDATE_DATA_KEY
from ap2.types.mandate import PaymentMandate
from common import artifact_utils
from common import message_utils
from common.a2a_extension_utils import EXTENSION_URI
from common.a2a_message_builder import A2aMessageBuilder
from common.payment_remote_a2a_client import PaymentRemoteA2aClient


async def initiate_payment(
    data_parts: list[dict[str, Any]],
    updater: TaskUpdater,
    current_task: Task | None,
    debug_mode: bool = False,
) -> None:
  """Handles the initiation of a payment."""
  payment_mandate = message_utils.find_data_part(
      PAYMENT_MANDATE_DATA_KEY, data_parts
  )
  if not payment_mandate:
    error_message = _create_text_parts("Missing payment_mandate.")
    await updater.failed(message=updater.new_agent_message(parts=error_message))
    return

  payment_mandate_obj = PaymentMandate.model_validate(payment_mandate)
  payment_method_type = payment_mandate_obj.payment_mandate_contents.payment_response.method_name

  # Route to appropriate handler based on payment method type
  if payment_method_type == "CARD":
    challenge_response = (
        message_utils.find_data_part("challenge_response", data_parts) or ""
    )
    await _handle_payment_mandate(
        payment_mandate_obj,
        challenge_response,
        updater,
        current_task,
        debug_mode,
    )
  elif payment_method_type == "BANK_ACCOUNT":
    await _handle_bank_payment(
        payment_mandate_obj,
        updater,
        current_task,
        debug_mode,
    )
  elif payment_method_type == "DIGITAL_WALLET":
    await _handle_digital_wallet_payment(
        payment_mandate_obj,
        updater,
        current_task,
        debug_mode,
    )
  else:
    error_message = _create_text_parts(f"Unsupported payment method: {payment_method_type}")
    await updater.failed(message=updater.new_agent_message(parts=error_message))


async def _handle_payment_mandate(
    payment_mandate: PaymentMandate,
    challenge_response: str,
    updater: TaskUpdater,
    current_task: Task | None,
    debug_mode: bool = False,
) -> None:
  """Handles a payment mandate.

  If no task is present, it initiates a transaction challenge. If a task
  requires input, it verifies the challenge response and completes the payment.

  Args:
    payment_mandate: The payment mandate containing payment details.
    challenge_response: The response to a transaction challenge, if any.
    updater: The task updater for managing task state.
    current_task: The current task, or None if it's a new payment.
    debug_mode: Whether the agent is in debug mode.
  """
  if current_task is None:
    await _raise_challenge(updater)
    return

  if current_task.status.state == TaskState.input_required:
    await _check_challenge_response_and_complete_payment(
        payment_mandate,
        challenge_response,
        updater,
        debug_mode,
    )
    return


async def _raise_challenge(
    updater: TaskUpdater,
) -> None:
  """Raises a transaction challenge.

  This challenge would normally be raised by the issuer, but we don't
  have an issuer in the demo, so we raise the challenge here. For concreteness,
  we are using an OTP challenge in this sample.

  Args:
    updater: The task updater.
  """
  challenge_data = {
      "type": "otp",
      "display_text": (
          "The payment method issuer sent a verification code to the phone "
          "number on file, please enter it below. It will be shared with the "
          "issuer so they can authorize the transaction."
          "(Demo only hint: the code is 123)"
      ),
  }
  text_part = TextPart(
      text="Please provide the challenge response to complete the payment."
  )
  data_part = DataPart(data={"challenge": challenge_data})
  message = updater.new_agent_message(
      parts=[Part(root=text_part), Part(root=data_part)]
  )
  await updater.requires_input(message=message)


async def _check_challenge_response_and_complete_payment(
    payment_mandate: PaymentMandate,
    challenge_response: str,
    updater: TaskUpdater,
    debug_mode: bool = False,
) -> None:
  """Checks the challenge response and completes the payment process.

  Checking the challenge response would be done by the issuer, but we don't
  have an issuer in the demo, so we do it here.

  Args:
    payment_mandate: The payment mandate.
    challenge_response: The challenge response.
    updater: The task updater.
    debug_mode: Whether the agent is in debug mode.
  """
  if _challenge_response_is_valid(challenge_response=challenge_response):
    await _complete_payment(payment_mandate, updater, debug_mode)
    return

  message = updater.new_agent_message(
      _create_text_parts("Challenge response incorrect.")
  )
  await updater.requires_input(message=message)


async def _complete_payment(
    payment_mandate: PaymentMandate,
    updater: TaskUpdater,
    debug_mode: bool = False,
) -> None:
  """Completes the payment process.

  Args:
    payment_mandate: The payment mandate.
    updater: The task updater.
    debug_mode: Whether the agent is in debug mode.
  """
  payment_mandate_id = (
      payment_mandate.payment_mandate_contents.payment_mandate_id
  )
  payment_credential = await _request_payment_credential(
      payment_mandate, updater, debug_mode
  )

  logging.info(
      "Calling issuer to complete payment for %s with payment credential %s...",
      payment_mandate_id,
      payment_credential,
  )
  # Call issuer to complete the payment
  success_message = updater.new_agent_message(
      parts=_create_text_parts("{'status': 'success'}")
  )
  await updater.complete(message=success_message)


def _challenge_response_is_valid(challenge_response: str) -> bool:
  """Validates the challenge response."""

  return challenge_response == "123"


async def _request_payment_credential(
    payment_mandate: PaymentMandate,
    updater: TaskUpdater,
    debug_mode: bool = False,
) -> str:
  """Sends a request to the Credentials Provider for payment credentials.

  Args:
    payment_mandate: The PaymentMandate containing payment details.
    updater: The task updater.
    debug_mode: Whether the agent is in debug mode.

  Returns:
    payment_credential: The payment credential details.
  """
  token_object = (
      payment_mandate.payment_mandate_contents.payment_response.details.get(
          "token"
      )
  )
  credentials_provider_url = token_object.get("url")

  credentials_provider = PaymentRemoteA2aClient(
      name="credentials_provider",
      base_url=credentials_provider_url,
      required_extensions={EXTENSION_URI},
  )

  message_builder = (
      A2aMessageBuilder()
      .set_context_id(updater.context_id)
      .add_text("Give me the payment method credentials for the given token.")
      .add_data(PAYMENT_MANDATE_DATA_KEY, payment_mandate.model_dump())
      .add_data("debug_mode", debug_mode)
  )
  task = await credentials_provider.send_a2a_message(message_builder.build())

  if not task.artifacts:
    raise ValueError("Failed to find the payment method data.")
  payment_credential = artifact_utils.get_first_data_part(task.artifacts)

  return payment_credential


async def _handle_bank_payment(
    payment_mandate: PaymentMandate,
    updater: TaskUpdater,
    current_task: Task | None,
    debug_mode: bool = False,
) -> None:
  """Handles bank account payment processing.

  Args:
    payment_mandate: The payment mandate containing bank account details.
    updater: The task updater for managing task state.
    current_task: The current task, or None if it's a new payment.
    debug_mode: Whether the agent is in debug mode.
  """
  if current_task is None:
    await _raise_bank_verification(updater)
    return

  if current_task.status.state == TaskState.input_required:
    await _verify_bank_account_and_complete_payment(
        payment_mandate,
        updater,
        debug_mode,
    )
    return


async def _handle_digital_wallet_payment(
    payment_mandate: PaymentMandate,
    updater: TaskUpdater,
    current_task: Task | None,
    debug_mode: bool = False,
) -> None:
  """Handles digital wallet payment processing.

  Args:
    payment_mandate: The payment mandate containing digital wallet details.
    updater: The task updater for managing task state.
    current_task: The current task, or None if it's a new payment.
    debug_mode: Whether the agent is in debug mode.
  """
  # Digital wallet payments are typically instant, but we should still validate
  # the wallet credentials before processing
  try:
    payment_method = await _request_payment_credential(payment_mandate, updater, debug_mode)

    # Validate digital wallet credentials
    if not payment_method or not payment_method.get("brand"):
      raise ValueError("Invalid digital wallet credentials")

    await _complete_digital_wallet_payment(payment_mandate, updater, debug_mode)
  except Exception as e:
    error_message = _create_text_parts(f"Digital wallet payment failed: {str(e)}")
    await updater.failed(message=updater.new_agent_message(parts=error_message))


async def _raise_bank_verification(updater: TaskUpdater) -> None:
  """Initiates bank account verification process.

  This function simulates the bank account verification flow by requesting
  user input for routing and account numbers. In a real implementation,
  this would integrate with banking APIs for verification.
  """
  verification_data = {
      "type": "bank_verification",
      "display_text": (
          "Please verify your bank account by entering the routing number "
          "and account number. (Demo: use routing '123456789' and account '111')"
      ),
  }
  text_part = TextPart(
      text="Please provide bank account verification to complete the payment."
  )
  data_part = DataPart(data={"verification": verification_data})
  message = updater.new_agent_message(
      parts=[Part(root=text_part), Part(root=data_part)]
  )
  await updater.requires_input(message=message)


async def _verify_bank_account_and_complete_payment(
    payment_mandate: PaymentMandate,
    updater: TaskUpdater,
    debug_mode: bool = False,
) -> None:
  """Verifies bank account and completes the payment."""
  # For demo purposes, we'll simulate successful verification
  # In a real implementation, this would validate against the bank account data
  # from the credentials provider

  # Get the bank account details from the payment mandate
  payment_method = await _request_payment_credential(payment_mandate, updater, debug_mode)

  # Demo verification logic - check if account number matches
  account_number = payment_method.get("account_number", "")

  if account_number == "111":  # This matches the demo bank account
    await _complete_bank_payment(payment_mandate, updater, debug_mode)
  else:
    error_message = _create_text_parts("Bank account verification failed. Invalid account number.")
    await updater.failed(message=updater.new_agent_message(parts=error_message))


async def _complete_bank_payment(
    payment_mandate: PaymentMandate,
    updater: TaskUpdater,
    debug_mode: bool = False,
) -> None:
  """Completes the bank payment processing."""
  payment_mandate_id = (
      payment_mandate.payment_mandate_contents.payment_mandate_id
  )

  logging.info(
      "Processing bank transfer for mandate %s...",
      payment_mandate_id,
  )

  # Simulate bank transfer processing
  transaction_id = f"bank_tx_{payment_mandate_id[-8:]}"
  success_message = updater.new_agent_message(
      parts=_create_text_parts(f"{{'status': 'success', 'method': 'bank_transfer', 'transaction_id': '{transaction_id}', 'mandate_id': '{payment_mandate_id}'}}")
  )
  await updater.complete(message=success_message)

  logging.info(
      "Bank transfer completed successfully for mandate %s, transaction %s",
      payment_mandate_id,
      transaction_id,
  )


async def _complete_digital_wallet_payment(
    payment_mandate: PaymentMandate,
    updater: TaskUpdater,
    debug_mode: bool = False,
) -> None:
  """Completes the digital wallet payment processing."""
  payment_mandate_id = (
      payment_mandate.payment_mandate_contents.payment_mandate_id
  )

  logging.info(
      "Processing digital wallet payment for mandate %s...",
      payment_mandate_id,
  )

  # Simulate instant digital wallet processing
  transaction_id = f"wallet_tx_{payment_mandate_id[-8:]}"
  success_message = updater.new_agent_message(
      parts=_create_text_parts(f"{{'status': 'success', 'method': 'digital_wallet', 'transaction_id': '{transaction_id}', 'mandate_id': '{payment_mandate_id}'}}")
  )
  await updater.complete(message=success_message)

  logging.info(
      "Digital wallet payment completed successfully for mandate %s, transaction %s",
      payment_mandate_id,
      transaction_id,
  )


def _create_text_parts(*texts: str) -> list[Part]:
  """Helper to create text parts."""
  return [Part(root=TextPart(text=text)) for text in texts]

File: samples/typescript/src/agents/merchant-payment-processor-agent/tools.ts
import { ai, z } from "./genkit.js";
import { A2AClient } from "@a2a-js/sdk/client";
import type {
  MessageSendParams,
  SendMessageSuccessResponse,
  Task,
  TaskStatusUpdateEvent,
} from "@a2a-js/sdk";
import { v4 as uuidv4 } from "uuid";
import type { ExecutionEventBus } from "@a2a-js/sdk/server";
import { findDataPart, parseCanonicalObject } from "../../utils/message.js";
import type { PaymentMandate } from "../../types/payment-mandate.js";
import { paymentMandateSchema } from "../../schemas/payment-mandate.js";

const PAYMENT_MANDATE_DATA_KEY = "ap2.mandates.PaymentMandate";

function challengeResponseIsValid(challengeResponse: string): boolean {
  return challengeResponse === "123";
}

async function requestPaymentCredential(
  paymentMandate: PaymentMandate,
  contextId: string,
  debugMode: boolean = false
): Promise<unknown> {
  const credentialsProviderUrl =
    "http://localhost:8002/.well-known/agent-card.json";

  const client = await A2AClient.fromCardUrl(credentialsProviderUrl);

  const message: MessageSendParams = {
    message: {
      messageId: uuidv4(),
      role: "user",
      contextId: contextId,
      parts: [
        {
          kind: "text",
          text: "Give me the payment method credentials for the given token.",
        },
        {
          kind: "data",
          data: {
            [PAYMENT_MANDATE_DATA_KEY]: paymentMandate,
          },
        },
        {
          kind: "data",
          data: {
            debug_mode: debugMode,
          },
        },
      ],
      kind: "message",
    },
  };

  const response = await client.sendMessage(message);

  if ("error" in response) {
    throw new Error(response.error.message);
  }

  const result = (response as SendMessageSuccessResponse).result;
  if (result.kind === "task") {
    const task = result as Task;
    if (!task.artifacts || task.artifacts.length === 0) {
      throw new Error("Failed to find the payment method data.");
    }

    const firstArtifact = task.artifacts[0];
    if (firstArtifact.parts && firstArtifact.parts.length > 0) {
      const dataPart = firstArtifact.parts.find((p) => p.kind === "data");
      if (dataPart && dataPart.kind === "data") {
        return dataPart.data;
      }
    }
  }

  throw new Error("Failed to retrieve payment credentials.");
}

async function raiseChallenge(
  eventBus: ExecutionEventBus,
  taskId: string,
  contextId: string
): Promise<void> {
  const challengeData = {
    type: "otp",
    display_text:
      "The payment method issuer sent a verification code to the phone " +
      "number on file, please enter it below. It will be shared with the " +
      "issuer so they can authorize the transaction." +
      "(Demo only hint: the code is 123)",
  };

  const statusUpdate: TaskStatusUpdateEvent = {
    kind: "status-update",
    taskId: taskId,
    contextId: contextId,
    status: {
      state: "input-required",
      message: {
        kind: "message",
        role: "agent",
        messageId: uuidv4(),
        parts: [
          {
            kind: "text",
            text: "Please provide the challenge response to complete the payment.",
          },
          {
            kind: "data",
            data: { challenge: challengeData },
          },
        ],
        taskId: taskId,
        contextId: contextId,
      },
      timestamp: new Date().toISOString(),
    },
    final: true,
  };

  eventBus.publish(statusUpdate);
}

async function completePayment(
  paymentMandate: PaymentMandate,
  eventBus: ExecutionEventBus,
  taskId: string,
  contextId: string,
  debugMode: boolean = false
): Promise<void> {
  await requestPaymentCredential(paymentMandate, contextId, debugMode);

  const successUpdate: TaskStatusUpdateEvent = {
    kind: "status-update",
    taskId: taskId,
    contextId: contextId,
    status: {
      state: "completed",
      message: {
        kind: "message",
        role: "agent",
        messageId: uuidv4(),
        parts: [
          {
            kind: "text",
            text: "{'status': 'success'}",
          },
        ],
        taskId: taskId,
        contextId: contextId,
      },
      timestamp: new Date().toISOString(),
    },
    final: true,
  };

  eventBus.publish(successUpdate);
}

export const initiatePayment = ai.defineTool(
  {
    name: "initiatePayment",
    description: "Initiates a payment for a given payment mandate.",
    inputSchema: z.object({
      dataParts: z
        .array(z.record(z.unknown()))
        .describe(
          "The data parts from the request, expected to contain a PaymentMandate and optionally a challenge response."
        ),
      eventBus: z
        .custom<ExecutionEventBus>()
        .describe("The event bus to add artifacts and complete the task."),
      currentTask: z.custom<Task>().optional(),
      taskId: z.string().optional().describe("The task ID from the executor"),
      contextId: z
        .string()
        .optional()
        .describe("The context ID from the executor"),
    }),
  },
  async (input) => {
    const { dataParts, eventBus, currentTask } = input;
    const taskId = input.taskId || currentTask?.id || uuidv4();
    const contextId = input.contextId || currentTask?.contextId || uuidv4();

    const paymentMandate = parseCanonicalObject<PaymentMandate>(
      PAYMENT_MANDATE_DATA_KEY,
      dataParts,
      paymentMandateSchema
    );

    if (!paymentMandate) {
      throw new Error("Missing payment_mandate.");
    }

    const challengeResponse = findDataPart("challenge_response", dataParts) as
      | string
      | null;
    const debugMode =
      (findDataPart("debug_mode", dataParts) as boolean | null) || false;

    if (!currentTask) {
      await raiseChallenge(eventBus, taskId, contextId);
      return { status: "input-required", message: "Challenge raised" };
    }

    if (currentTask.status.state === "input-required") {
      if (!challengeResponse) {
        throw new Error("Challenge response is required.");
      }

      if (!challengeResponseIsValid(challengeResponse)) {
        const statusUpdate: TaskStatusUpdateEvent = {
          kind: "status-update",
          taskId: taskId,
          contextId: contextId,
          status: {
            state: "input-required",
            message: {
              kind: "message",
              role: "agent",
              messageId: uuidv4(),
              parts: [
                {
                  kind: "text",
                  text: "Challenge response incorrect.",
                },
              ],
              taskId: taskId,
              contextId: contextId,
            },
            timestamp: new Date().toISOString(),
          },
          final: false,
        };
        eventBus.publish(statusUpdate);
        return { status: "input-required", message: "Invalid challenge" };
      }

      await completePayment(
        paymentMandate,
        eventBus,
        taskId,
        contextId,
        debugMode
      );
      return { status: "completed", message: "Payment completed" };
    }

    return { status: "unknown", message: "Unexpected task state" };
  }
);
